In [ ]:
Kiwi pythonweekend, cesta na devs

## Requests - knihovna pro komunikování s  API a získávání informací z webu (ang. scraping)


In [1]:
import requests

*IMDB je obdoba csfd.cz - dazabaze filmů

In [2]:
requests.get("https://www.imdb.com/chart/top/")

<Response [200]>

In [9]:
response = requests.get("https://www.imdb.com/chart/top/")

In [4]:
print(response.text)




<!DOCTYPE html>
<html
    xmlns:og="http://ogp.me/ns#"
    xmlns:fb="http://www.facebook.com/2008/fbml">
    <head>
         

        <meta charset="utf-8">
        <meta http-equiv="X-UA-Compatible" content="IE=edge">

    
    
    

    
    
    




        <script type="text/javascript">var IMDbTimer={starttime: new Date().getTime(),pt:'java'};</script>

<script>
    if (typeof uet == 'function') {
      uet("bb", "LoadTitle", {wb: 1});
    }
</script>
  <script>(function(t){ (t.events = t.events || {})["csm_head_pre_title"] = new Date().getTime(); })(IMDbTimer);</script>
        <title>Top 250 Movies - IMDb</title>
  <script>(function(t){ (t.events = t.events || {})["csm_head_post_title"] = new Date().getTime(); })(IMDbTimer);</script>
<script>
    if (typeof uet == 'function') {
      uet("be", "LoadTitle", {wb: 1});
    }
</script>
<script>
    if (typeof uex == 'function') {
      uex("ld", "LoadTitle", {wb: 1});
    }
</script>

        <link rel="canonical" href="https:

In [8]:
#např: regiojet
# www.regiojet.cz
regiojet_lokace_z_api = requests.get("https://brn-ybus-pubapi.sa.cz/restapi/consts/locations")
regiojet_lokace = regiojet_lokace_z_api.json()
regiojet_lokace
for country in regiojet_lokace:
       print(country["country"])

[{'country': 'Germany',
  'code': 'DE',
  'cities': [{'id': 10202072,
    'name': 'Berlin',
    'aliases': ['柏林', 'Берлин', 'Berlín', 'Берлін'],
    'stationsTypes': ['BUS_STATION'],
    'stations': [{'id': 2704917000,
      'name': 'Südkreuz Bf',
      'fullname': 'Berlin - Südkreuz Bf',
      'aliases': [],
      'address': 'Berlin Südkreuz, Hildegard-Knef-Platz, 10829 Berlin, Německo',
      'stationsTypes': ['BUS_STATION'],
      'iataCode': 'BER',
      'stationUrl': 'https://www.regiojet.com/timetables-and-stops/stops/2704917000.html',
      'stationTimeZoneCode': 'Europe/Berlin',
      'wheelChairPlatform': None,
      'significance': 1,
      'longitude': 13.364163398742676,
      'latitude': 52.476200103759766,
      'imageUrl': None},
     {'id': 10204076,
      'name': 'ZOB am Funkturm',
      'fullname': 'Berlin - ZOB am Funkturm',
      'aliases': [],
      'address': 'ZOB Berlin, Hardenbergstraße 14, 10623 Berlin, Německo',
      'stationsTypes': ['BUS_STATION'],
      'i

### Status kody 

#### pravidlo palce: 2xx - dobrý (200 nejlepší :), ostatní - špatný (4xx velký špatný)

##### úkol: najděte co znamená status kod 418 - https://cs.wikipedia.org/wiki/Stavov%C3%A9_k%C3%B3dy_HTTP

### Headers

In [11]:
from pprint import pprint

In [12]:
pprint(response.headers)

{'Content-Type': 'text/html;charset=UTF-8', 'Transfer-Encoding': 'chunked', 'Connection': 'keep-alive', 'Server': 'Server', 'Date': 'Sat, 27 Nov 2021 08:33:25 GMT', 'x-amz-rid': '7G38387Y3Z6PPQG6H573', 'Set-Cookie': 'uu=eyJpZCI6InV1MTRiNzBiMzNlNmExNGI5NTkyZTkiLCJwcmVmZXJlbmNlcyI6eyJmaW5kX2luY2x1ZGVfYWR1bHQiOmZhbHNlfX0=; Domain=.imdb.com; Expires=Thu, 15-Dec-2089 11:47:32 GMT; Path=/; Secure, session-id=000-0000000-0000000; Domain=.imdb.com; Expires=Thu, 15-Dec-2089 11:47:32 GMT; Path=/; Secure, session-id-time=2268722005; Domain=.imdb.com; Expires=Thu, 15-Dec-2089 11:47:32 GMT; Path=/; Secure', 'X-Frame-Options': 'SAMEORIGIN', 'Content-Security-Policy': "frame-ancestors 'self' imdb.com *.imdb.com *.media-imdb.com withoutabox.com *.withoutabox.com amazon.com *.amazon.com amazon.co.uk *.amazon.co.uk amazon.de *.amazon.de translate.google.com images.google.com www.google.com www.google.co.uk search.aol.com bing.com www.bing.com", 'Content-Language': 'en-US', 'Strict-Transport-Security': '

In [13]:
for el, value in response.headers.items():
    print(el + " : " + value)

Content-Type : text/html;charset=UTF-8
Transfer-Encoding : chunked
Connection : keep-alive
Server : Server
Date : Sat, 27 Nov 2021 08:33:25 GMT
x-amz-rid : 7G38387Y3Z6PPQG6H573
Set-Cookie : uu=eyJpZCI6InV1MTRiNzBiMzNlNmExNGI5NTkyZTkiLCJwcmVmZXJlbmNlcyI6eyJmaW5kX2luY2x1ZGVfYWR1bHQiOmZhbHNlfX0=; Domain=.imdb.com; Expires=Thu, 15-Dec-2089 11:47:32 GMT; Path=/; Secure, session-id=000-0000000-0000000; Domain=.imdb.com; Expires=Thu, 15-Dec-2089 11:47:32 GMT; Path=/; Secure, session-id-time=2268722005; Domain=.imdb.com; Expires=Thu, 15-Dec-2089 11:47:32 GMT; Path=/; Secure
X-Frame-Options : SAMEORIGIN
Content-Security-Policy : frame-ancestors 'self' imdb.com *.imdb.com *.media-imdb.com withoutabox.com *.withoutabox.com amazon.com *.amazon.com amazon.co.uk *.amazon.co.uk amazon.de *.amazon.de translate.google.com images.google.com www.google.com www.google.co.uk search.aol.com bing.com www.bing.com
Content-Language : en-US
Strict-Transport-Security : max-age=47474747; includeSubDomains; preloa

Headers se používají na specifikaci - encoding, content-type, tokeny, footprint zařízení a další

### Parameters

## API vs web

**scrapování webu = dostat data přímo z webových stránek**
- ideální pro osobní projekty, automatizaci - hlídací psi atd.
- "unikátní" projekty, nestandartní data 
- firmy scrapují konkurenci, Kiwi scrapuje data o cenách letenek

#### problémy scrapování
- diskuze legální/nelegální
- zatěžování serverů
- nestandartizované 
- snadná porucha - jakákoliv změna webu

## API - Application Programming Interface

trocha teorie z wiki:
Jde o sbírku procedur, funkcí, tříd či protokolů nějaké knihovny (ale třeba i jiného programu). API určuje, jakým způsobem jsou funkce knihovny volány ze zdrojového kódu programu. **Funkce API jsou programové celky, které programátor používá namísto toho, aby je sám naprogramoval.** https://cs.wikipedia.org/wiki/API

API is the acronym for Application Programming Interface, which is a software intermediary that **allows two applications to talk to each other. Each time you use an app like Facebook, send an instant message, or check the weather on your phone, you're using an API.**

### API je všude

abychom nemuseli "znovu vynalézat kolo" používáme API které už vytvořili ostatní programátoři nebo i firmy samotné (aby se vyhly scrapingu :) 

##### API je byznys

##### Příklady API:
    - počasí - https://openweathermap.org/current
    - burzovní info: https://yahooquery.dpguthrie.com/, https://yahooquery-streamlit.herokuapp.com/
    - státní svátky: https://calendarific.com/api/v2/holidays, https://calendarific.com/api-documentation
    - převádění měnových kurzů
    - pokeapi
    - yoda translator - https://funtranslations.com/yoda

#### Tvorba vlastních API - python frameworks Flask, FastAPI

## API klíč - API Key - api_key

**nutný pro autentizaci, z pravidla generované pro každého uživatele unikátně**

#### společný úkol: vytvořmě překladač do yodovštiny
https://funtranslations.com/api/yoda

Max 5 requestu za hodinu

Příklad

In [ ]:
# link https://www.theguardian.com/world/2011/apr/13/czech-president-steals-pen

In [ ]:
text = "As Piñera addresses journalists and gives Klaus an enthusiastic welcome, the Czech president takes the pen out, examines it carefully, moves his hands under the table, shuffles with his jacket, then buttons it up with both hands and finally lets his empty hands emerge above the table again to close the case."

yoda_speak = requests.get(f"https://api.funtranslations.com/translate/yoda.json?text={text}")
print(type(yoda_speak.json()))

In [ ]:
url = "https://api.funtranslations.com/translate/yoda.json?"
parametry = {"text" : text}
yoda_1 = requests.get(url=url, params=parametry)
print(yoda_1.json())

## Uchovávání hesel mimo kód - python-decouple

Není bezpečné uchovávat hesla, username, API klíče a další důležité přístupové věci v kódu. Strandardem je uchovávat je v `.env` souboru, který je "neviditelný". `.env` soubor vytvoříme v repozitáři kodu. Můžeme použít command line v terminálu

#### decouple knihovna

### úkol1 - Napiště funkci překladač,  která překládá pomocí deepL API

deepl API docs: https://www.deepl.com/docs-api

API klíč = "224dd4fb-9c29-8ee9-759a-a71057134c5d:fx" limit 500,000 znaků za měsíc

In [ ]:
def translate(text, target_lang, source_lang=None):
    pass

### úkol2 - přidejte funkci translator která se zeptá uživatele na text a do jakého jazyku ho chce přeložit

In [ ]:
def translator():
    pass

### úkol3 - přidejte možnost odeslat překlad na email

In [ ]:
def send_email(receiver_email, message):
    pass

### TODO
- selenium - ukázka
- bs4 praktická ukázka
- regex


## Scraping with BeautifulSoup

- IMDB - https://www.imdb.com/chart/top/
- bs4 - https://beautiful-soup-4.readthedocs.io/en/latest/

BeautifulSoup(=bs) object reprezentuje zparsovaný **html dokument = webovou stránku**. Můžeme parsovat i xml. Můžeme si to představit jako webovou stránku rozloženou do elementů (tagů), který má každý svoji třídu nebo id. Beautiful soup nám potom dovoluje procházet tento soubor, hledat v něm a ukazovat data

##### syntax

In [14]:
import requests
from bs4 import BeautifulSoup

r = requests.get("https://www.imdb.com/chart/top/")
print(r.text)




<!DOCTYPE html>
<html
    xmlns:og="http://ogp.me/ns#"
    xmlns:fb="http://www.facebook.com/2008/fbml">
    <head>
         

        <meta charset="utf-8">
        <meta http-equiv="X-UA-Compatible" content="IE=edge">

    
    
    

    
    
    




        <script type="text/javascript">var IMDbTimer={starttime: new Date().getTime(),pt:'java'};</script>

<script>
    if (typeof uet == 'function') {
      uet("bb", "LoadTitle", {wb: 1});
    }
</script>
  <script>(function(t){ (t.events = t.events || {})["csm_head_pre_title"] = new Date().getTime(); })(IMDbTimer);</script>
        <title>Top 250 Movies - IMDb</title>
  <script>(function(t){ (t.events = t.events || {})["csm_head_post_title"] = new Date().getTime(); })(IMDbTimer);</script>
<script>
    if (typeof uet == 'function') {
      uet("be", "LoadTitle", {wb: 1});
    }
</script>
<script>
    if (typeof uex == 'function') {
      uex("ld", "LoadTitle", {wb: 1});
    }
</script>

        <link rel="canonical" href="https:

In [16]:
import requests
from bs4 import BeautifulSoup

r = requests.get("https://www.imdb.com/chart/top/")
soup = BeautifulSoup(r.text, "html.parser")
print(type(soup))

<class 'bs4.BeautifulSoup'>


### html & css elementy

https://www.techonthenet.com/html/elements/tr_tag.php všechny typy tagu

### BS metoda - find_all() a find()

In [ ]:
# syntax
all_tags = soup.find_all("tag", class_="class_name")
# vrací objekt (něco jako list) přes který se dá iterovat
for tag in all_tags:
    tag.find("sub_tag").text
    tag.find("sub_tag").get("key")

Nejjednodušším způsobem jak navigovat přes zparsovaný html dokument je přes tagy a jejich **id** a **class**

tips & tricks https://hackersandslackers.com/scraping-urls-with-beautifulsoup/

In [23]:
r = requests.get("https://www.imdb.com/chart/top/")
soup = BeautifulSoup(r.text, "html.parser")
filmy = soup.find_all("td", class_="titleColumn")
print(filmy[0])

<td class="titleColumn">
      1.
      <a href="/title/tt0111161/" title="Frank Darabont (dir.), Tim Robbins, Morgan Freeman">Vykoupení z věznice Shawshank</a>
<span class="secondaryInfo">(1994)</span>
</td>


### úkol stáhnetě z webu imdb top 250 filmů, jejich rok vydání. seřaďte je podle data vydání od nejnovějšího

In [27]:
import requests
from bs4 import BeautifulSoup

r = requests.get("https://www.imdb.com/chart/top/")
r.headers
soup = BeautifulSoup(r.text, "html.parser")
filmy = soup.find_all("td", class_="titleColumn")
print(r.headers)

{'Content-Type': 'text/html;charset=UTF-8', 'Transfer-Encoding': 'chunked', 'Connection': 'keep-alive', 'Server': 'Server', 'Date': 'Sat, 27 Nov 2021 12:41:28 GMT', 'x-amz-rid': '1J2VY5Q3ZCGQS90KJYQS', 'Set-Cookie': 'uu=eyJpZCI6InV1NDVlMmI3MzMyYTY1NDM3MTg2MWEiLCJwcmVmZXJlbmNlcyI6eyJmaW5kX2luY2x1ZGVfYWR1bHQiOmZhbHNlfX0=; Domain=.imdb.com; Expires=Thu, 15-Dec-2089 15:55:35 GMT; Path=/; Secure, session-id=000-0000000-0000000; Domain=.imdb.com; Expires=Thu, 15-Dec-2089 15:55:35 GMT; Path=/; Secure, session-id-time=2268736888; Domain=.imdb.com; Expires=Thu, 15-Dec-2089 15:55:35 GMT; Path=/; Secure', 'X-Frame-Options': 'SAMEORIGIN', 'Content-Security-Policy': "frame-ancestors 'self' imdb.com *.imdb.com *.media-imdb.com withoutabox.com *.withoutabox.com amazon.com *.amazon.com amazon.co.uk *.amazon.co.uk amazon.de *.amazon.de translate.google.com images.google.com www.google.com www.google.co.uk search.aol.com bing.com www.bing.com", 'Content-Language': 'en-US', 'Strict-Transport-Security': '

### bonus - stáhněte informace o režisérovi a hercích z odkazu. Uložte je do csv souboru jako imdb_top250.csv

In [25]:
reziser_a_herci = film.find("a").get("title")
x = reziser_a_herci.strip("(dir.)")
x

'Shoojit Sircar (dir.), Vicky Kaushal, Banita Sandhu'

In [33]:
rah = 'Shoojit Sircar (dir.), Vicky Kaushal, Banita Sandhu'
rah_list = rah.split(",")
#print(rah_list)
reziser, *herci= rah_list
print(reziser)
print(herci)

Shoojit Sircar (dir.)
[' Vicky Kaushal', ' Banita Sandhu']


In [38]:
reziser = "dirShoojit Sircar (dir.)"
rez = reziser[0:-6]

Shoojit Sircar 


In [ ]:
reziser_a_herci = [film.find("a").get("title").split(",")[0].strip(" (dir.)")]+[film.find("a").get("title").split(",")[1:]]

In [ ]:
list_cisel = []
for i in range(20):
    list_cisel.append(i)


In [41]:
list_cisel = [i for i in range(1,21)]
list_cisel

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]